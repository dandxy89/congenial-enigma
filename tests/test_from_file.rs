use std::path::PathBuf;

use lp_parser_rs::{
    model::{lp_error::LPParserError, lp_problem::LPProblem},
    parse::{parse_file, parse_lp_file},
};

#[macro_export]
macro_rules! generate_test {
    ($test_name:ident, $file:expr, $o_sum:expr, $c_sum:expr) => {
        #[test]
        fn $test_name() {
            let result = read_file_from_resources($file).unwrap();

            #[cfg(feature = "serde")]
            insta::assert_yaml_snapshot!(result, {
                ".variables" => insta::sorted_redaction(),
                ".objectives" => insta::sorted_redaction(),
                ".constraints" => insta::sorted_redaction(),
            });

            let summation = &result.objectives.iter().map(|o| o.coefficients.iter().map(|c| c.coefficient).sum::<f64>()).sum();
            float_eq::assert_float_eq!($o_sum, *summation, abs <= 1e-2);

            let summation = &result.constraints.iter().map(|(_, c)| c.coefficients().iter().map(|c| c.coefficient).sum::<f64>()).sum();
            float_eq::assert_float_eq!($c_sum, *summation, abs <= 1e-2);
        }
    };
}

generate_test!(afiro, "afiro.lp", -4.100, 25.369);
generate_test!(afiro_ext, "afiro_ext.lp", -5.239, 25.369);
generate_test!(boeing1, "boeing1.lp", 1187.985, 194612.346);
generate_test!(boeing2, "boeing2.lp", 78.488, 20863.836);
generate_test!(fit1d, "fit1d.lp", 82457.0, -146871.180);
generate_test!(kb2, "kb2.lp", 11.675, 10143.724);
generate_test!(pulp, "pulp.lp", 1., 73.44);
generate_test!(pulp2, "pulp2.lp", -6.0, 147.);
generate_test!(sc50a, "sc50a.lp", -1., 30.3);
generate_test!(model2, "model2.lp", 0., 6.);
generate_test!(limbo, "limbo.lp", 2., 0.);
generate_test!(obj3_2cons, "3obj_2cons.lp", 1., -186.);
generate_test!(obj_2cons_only_binary_vars, "2obj_2cons_only_binary_vars.lp", -7.5, -186.);
generate_test!(obj_2cons_all_variable_types, "2obj_2cons_all_variable_types.lp", -7.5, -186.);
generate_test!(obj_1cons_all_variables_with_bounds, "1obj_1cons_all_variables_with_bounds.lp", -1., 16.5);
generate_test!(semi_continuous, "semi_continuous.lp", 2., 0.);
generate_test!(sos, "sos.lp", 0., 17.5);
generate_test!(test, "test.lp", 2., 2.9899);
generate_test!(test2, "test2.lp", -6., 147.);
generate_test!(empty_bounds, "empty_bounds.lp", 11., 2.);
generate_test!(blank_lines, "blank_lines.lp", 11., 2.);
generate_test!(optional_labels, "optional_labels.lp", 11., 1.);
generate_test!(infile_comments, "infile_comments.lp", 43., 7.);
generate_test!(infile_comments2, "infile_comments2.lp", 43., 0.);
generate_test!(missing_signs, "missing_signs.lp", 43., -3.);
generate_test!(fit2d, "fit2d.lp", 349048.9, -296677.389);
generate_test!(scientific_notation, "scientific_notation.lp", 238.0324, -43.87680);

#[test]
fn invalid() {
    let result = read_file_from_resources("invalid.lp");
    assert!(result.is_err());
}

fn read_file_from_resources(file_name: &str) -> Result<LPProblem, LPParserError> {
    let mut file_path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
    file_path.push(format!("resources/{file_name}"));
    let contents = parse_file(&file_path)?;
    parse_lp_file(&contents)
}
