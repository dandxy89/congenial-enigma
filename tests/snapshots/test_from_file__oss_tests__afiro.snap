---
source: tests/test_from_file.rs
expression: result
---
constraints:
  "R09:":
    Standard:
      name: "R09:"
      coefficients:
        - coefficient: -1
          var_name: X01
        - coefficient: 1
          var_name: X02
        - coefficient: 1
          var_name: X03
      sense: "="
      rhs: 0
  "R10:":
    Standard:
      name: "R10:"
      coefficients:
        - coefficient: -1.06
          var_name: X01
        - coefficient: 1
          var_name: X04
      sense: "="
      rhs: 0
  "R12:":
    Standard:
      name: "R12:"
      coefficients:
        - coefficient: -1
          var_name: X06
        - coefficient: -1
          var_name: X07
        - coefficient: -1
          var_name: X08
        - coefficient: -1
          var_name: X09
        - coefficient: 1
          var_name: X14
        - coefficient: 1
          var_name: X15
      sense: "="
      rhs: 0
  "R13:":
    Standard:
      name: "R13:"
      coefficients:
        - coefficient: -1.06
          var_name: X06
        - coefficient: -1.06
          var_name: X07
        - coefficient: -0.96
          var_name: X08
        - coefficient: -0.86
          var_name: X09
        - coefficient: 1
          var_name: X16
      sense: "="
      rhs: 0
  "R19:":
    Standard:
      name: "R19:"
      coefficients:
        - coefficient: -1
          var_name: X22
        - coefficient: 1
          var_name: X23
        - coefficient: 1
          var_name: X24
        - coefficient: 1
          var_name: X25
      sense: "="
      rhs: 0
  "R20:":
    Standard:
      name: "R20:"
      coefficients:
        - coefficient: -0.43
          var_name: X22
        - coefficient: 1
          var_name: X26
      sense: "="
      rhs: 0
  "R22:":
    Standard:
      name: "R22:"
      coefficients:
        - coefficient: -0.43
          var_name: X28
        - coefficient: -0.43
          var_name: X29
        - coefficient: -0.39
          var_name: X30
        - coefficient: -0.37
          var_name: X31
        - coefficient: 1
          var_name: X38
      sense: "="
      rhs: 0
  "R23:":
    Standard:
      name: "R23:"
      coefficients:
        - coefficient: 1
          var_name: X28
        - coefficient: 1
          var_name: X29
        - coefficient: 1
          var_name: X30
        - coefficient: 1
          var_name: X31
        - coefficient: -1
          var_name: X36
        - coefficient: 1
          var_name: X37
        - coefficient: 1
          var_name: X39
      sense: "="
      rhs: 44
  "X05:":
    Standard:
      name: "X05:"
      coefficients:
        - coefficient: 1
          var_name: X01
      sense: "<="
      rhs: 80
  "X17:":
    Standard:
      name: "X17:"
      coefficients:
        - coefficient: 1
          var_name: X06
        - coefficient: -1
          var_name: X10
      sense: "<="
      rhs: 80
  "X18:":
    Standard:
      name: "X18:"
      coefficients:
        - coefficient: 1
          var_name: X07
        - coefficient: -1
          var_name: X11
      sense: "<="
      rhs: 0
  "X19:":
    Standard:
      name: "X19:"
      coefficients:
        - coefficient: 1
          var_name: X08
        - coefficient: -1
          var_name: X12
      sense: "<="
      rhs: 0
  "X20:":
    Standard:
      name: "X20:"
      coefficients:
        - coefficient: 1
          var_name: X09
        - coefficient: -1
          var_name: X13
      sense: "<="
      rhs: 0
  "X21:":
    Standard:
      name: "X21:"
      coefficients:
        - coefficient: -1
          var_name: X02
        - coefficient: 1.4
          var_name: X14
      sense: "<="
      rhs: 0
  "X27:":
    Standard:
      name: "X27:"
      coefficients:
        - coefficient: 1
          var_name: X22
      sense: "<="
      rhs: 500
  "X40:":
    Standard:
      name: "X40:"
      coefficients:
        - coefficient: 1
          var_name: X28
        - coefficient: -1
          var_name: X32
      sense: "<="
      rhs: 500
  "X41:":
    Standard:
      name: "X41:"
      coefficients:
        - coefficient: 1
          var_name: X29
        - coefficient: -1
          var_name: X33
      sense: "<="
      rhs: 0
  "X42:":
    Standard:
      name: "X42:"
      coefficients:
        - coefficient: 1
          var_name: X30
        - coefficient: -1
          var_name: X34
      sense: "<="
      rhs: 0
  "X43:":
    Standard:
      name: "X43:"
      coefficients:
        - coefficient: 1
          var_name: X31
        - coefficient: -1
          var_name: X35
      sense: "<="
      rhs: 0
  "X44:":
    Standard:
      name: "X44:"
      coefficients:
        - coefficient: -1
          var_name: X23
        - coefficient: 1.4
          var_name: X36
      sense: "<="
      rhs: 0
  "X45:":
    Standard:
      name: "X45:"
      coefficients:
        - coefficient: 2.364
          var_name: X10
        - coefficient: 2.386
          var_name: X11
        - coefficient: 2.408
          var_name: X12
        - coefficient: 2.429
          var_name: X13
        - coefficient: -1
          var_name: X25
        - coefficient: 2.191
          var_name: X32
        - coefficient: 2.219
          var_name: X33
        - coefficient: 2.249
          var_name: X34
        - coefficient: 2.279
          var_name: X35
      sense: "<="
      rhs: 0
  "X46:":
    Standard:
      name: "X46:"
      coefficients:
        - coefficient: -1
          var_name: X03
        - coefficient: 0.109
          var_name: X22
      sense: "<="
      rhs: 0
  "X47:":
    Standard:
      name: "X47:"
      coefficients:
        - coefficient: -1
          var_name: X15
        - coefficient: 0.109
          var_name: X28
        - coefficient: 0.108
          var_name: X29
        - coefficient: 0.108
          var_name: X30
        - coefficient: 0.107
          var_name: X31
      sense: "<="
      rhs: 0
  "X48:":
    Standard:
      name: "X48:"
      coefficients:
        - coefficient: 0.301
          var_name: X01
        - coefficient: -1
          var_name: X24
      sense: "<="
      rhs: 0
  "X49:":
    Standard:
      name: "X49:"
      coefficients:
        - coefficient: 0.301
          var_name: X06
        - coefficient: 0.313
          var_name: X07
        - coefficient: 0.313
          var_name: X08
        - coefficient: 0.326
          var_name: X09
        - coefficient: -1
          var_name: X37
      sense: "<="
      rhs: 0
  "X50:":
    Standard:
      name: "X50:"
      coefficients:
        - coefficient: 1
          var_name: X04
        - coefficient: 1
          var_name: X26
      sense: "<="
      rhs: 310
  "X51:":
    Standard:
      name: "X51:"
      coefficients:
        - coefficient: 1
          var_name: X16
        - coefficient: 1
          var_name: X38
      sense: "<="
      rhs: 300
objectives:
  - coefficients:
      - coefficient: -10.4
        var_name: X03
      - coefficient: -2.5
        var_name: X07
      - coefficient: 2.1
        var_name: X10
    name: OBJ3
  - coefficients:
      - coefficient: -0.4
        var_name: X02
      - coefficient: -0.32
        var_name: X14
      - coefficient: -0.6
        var_name: X23
      - coefficient: -0.48
        var_name: X36
      - coefficient: 10
        var_name: X39
    name: OBJ1
  - coefficients:
      - coefficient: 1
        var_name: X03
      - coefficient: 2.5
        var_name: X06
      - coefficient: -5
        var_name: X08
    name: OBJ2
problem_name: afiro.mps
problem_sense: Minimize
variables:
  X01: General
  X02: General
  X03: General
  X04: General
  X06: General
  X07: General
  X08: General
  X09: General
  X10: General
  X11: General
  X12: General
  X13: General
  X14: General
  X15: General
  X16: General
  X22: General
  X23: General
  X24: General
  X25: General
  X26: General
  X28: General
  X29: General
  X30: General
  X31: General
  X32: General
  X33: General
  X34: General
  X35: General
  X36: General
  X37: General
  X38: General
  X39: General
